### ISSUE TITLE
Support for variable substitution when using docker-compose.yaml

### ISSUE THREAD

**Issue by joshua-barnett** [microsoft/vscode-dev-containers#1385]
I am trying to integrate the [Remote - Containers](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) with my `docker-compose.yaml`.

However, my compose file is making use of the [variable substitution](https://docs.docker.com/compose/compose-file/compose-file-v3/#variable-substitution) feature.

```yaml
services:
    command: yarn server
    environment:
        MONGO_URI: mongodb://database:27017/todo
    ports:
        - ${SERVER_PORT}:3000
        - ${SERVER_DEBUG_PORT}:9320
```

Typically I set these variables from a `Makefile` that wraps the `docker-compose` commands I run to start my development environment.

### How can I set these environment variables before the extension creates the **Dev Container** via `docker-compose`?

I've attempted to use the `initializeCommand` in the following ways.

`devcontainer.json`

```json
"initializeCommand": ". ./env.sh"
```

```json
"initializeCommand": "source env.sh"
```

I also attempted to use `remoteEnv` in the `devcontainer.json`.

I also attempted to set `settings.terminal.integrated.env.linux`.

**Comment by Chuxel** [comment]
Docker Compose's model here would be to put a .env file next to your Docker Compose file with the variables you'd like to use in it. 

> You can set default values for environment variables using a [.env file](https://docs.docker.com/compose/env-file/), which Compose automatically looks for in project directory (parent folder of your Compose file). Values set in the shell environment override those set in the .env file.

https://docs.docker.com/compose/environment-variables/

Specifying `env_file:` for a given service/container then allows you to pass variables from an env file into the container, but just putting a .env file in the right spot will cause it to be picked up inside the compose file itself.

Hope that helps!

**Comment by divmgl** [comment]
That's the point. We want to be able to specify the `env_file`.

**Comment by Chuxel** [comment]
@divmgl If you put `.env` next to your Docker Compose file, it will be pulled in without specifying it. You can then use those variables in `environment` or `ports` or anywhere else.  initalizeCommand can also create that file if needed.

**Comment by Chuxel** [comment]
Yeah unfortunately the alternative is to use `initalizeCommand` to concatenate contents to the `.env` file, but I understand this could cause it to get inadvertently committed.  Looks like there's a couple issues out there with various forms of wanting docker compose up argument (https://github.com/microsoft/vscode-remote-release/issues/6298 is a similar one to the one it looks like you raised).

**Comment by sliekens** [comment]
I also found myself in a position where I want to do something with environment variables.

I want to mount dotnet user-secrets from the host machine inside my dev container. The path of the secrets is platform-specific so I can't use a constant, it has to be a variable because I don't want to tie my dev container to one OS.

``` yml
# .devcontainer/docker-compose.yml
version: "3.7"
services:
  workspace:
    user: vscode
    volumes:
      - ${USERSECRETS}:/home/vscode/.microsoft/usersecrets:cached
```

The `USERSECRETS` is not an environment variable that already exists so I tried to create it with an initializeCommand.

.devcontainer/devcontainer.json

``` json
{
  "initializeCommand": [
    ".devcontainer/initializeHost"
  ],
}
```

Mac/Linux

``` sh
# .devcontainer/initializeHost
export USERSECRETS=$HOME/.microsoft/usersecrets
```

Windows

``` bat
REM .devcontainer/initializeHost.cmd
set USERSECRETS %APPDATA%/Microsoft/UserSecrets
```

So as you already know, this doesn't work and it's not clear to me what I should try next. I'm not really interested in using `.env` files except as a hack.

**Comment by liiight** [comment]
I think this should not be closed as it is appeardit was solved, and it wasn't. Please consider reopening it.
